[Train] Add example of fine-tuning Llama-2 on Intel Gaudi

harborn opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

To leverage the potential of Intel Gaudi accelerator, we extend Ray Train's capabilities by adding support for Intel Gaudi (HPU) hardware. This PR include an example for fine-tuning Llama-2-7b on multi HPUs.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

jerome-habana commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
```diff
@@ -0,0 +1,771 @@
+    "\n",
+    "In this Jupyter notebook, we will fine-tune a [Llama-2-7b](https://huggingface.co/meta-llama/Llama-2-7b-chat-hf) model by using HPU in DDP accelerate mode. We will use PyTorch for model training and Ray for distributed training. We will use dataset [tatsu-lab/alpaca](https://huggingface.co/datasets/tatsu-lab/alpaca).\n",
+    "\n",
+    "[Habana Gaudi AI Processors (HPUs)](https://habana.ai) are AI hardware accelerators designed by Habana Labs. For more information, see [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/).\n",
```
Intel Gaudi

aslonnie reviewed (commented):
let docs reviewer know when this is ready for review.

woshiyyya commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
```diff
@@ -0,0 +1,831 @@
+    "\n",
+    "def train_llama(num_workers, execution_mode):\n",
+    "    # Setting environment variables\n",
+    "    os.environ[\"RAY_EXPERIMENTAL_NOSET_HABANA_VISIBLE_MODULES\"] = \"true\"\n",
```
Why are we setting this env var here?

harborn replied:
maybe should remove, I will try.

harborn replied:
removed those env var

woshiyyya commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
```diff
@@ -0,0 +1,831 @@
+       "    <div class=\"vDivider\"></div>\n",
+       "    <div class=\"systemInfo\">\n",
+       "      <h3>System Info</h3>\n",
+       "      Using FIFO scheduling algorithm.<br>Logical resource usage: 5.0/160 CPUs, 0/0 GPUs (0.0/1.0 TPU, 4.0/8.0 HPU)\n",
```
<img width="792" alt="Screenshot 2024-05-10 at 1 16 30 PM" src="https://github.com/ray-project/ray/assets/26745457/b9ce44ee-c9a4-4eb1-9648-b83d7313a23f">

Nit: saw there is a TPU in your ray cluster. Is it colocated with HPUs on the same host? Or it's just another TPU node.

harborn replied:
Maybe this is ray's device detection bug I think.

woshiyyya replied:
Got it. Let me report this bug to Ray Core team.

woshiyyya commented on doc/source/train/examples/intel_gaudi/llama_deepspeed.ipynb:
```diff
@@ -0,0 +1,2436 @@
+    "\n",
+    "    runtime_env = {\n",
+    "        \"env_vars\": {\n",
+    "            \"PT_HPU_MAX_COMPOUND_OP_SIZE\": \"10\",\n",
+    "            \"DEEPSPEED_HPU_ZERO3_SYNC_MARK_STEP_REQUIRED\": \"1\",\n",
+    "            \"WORLD_SIZE\": str(num_workers),\n",
+    "            \"HABANA_VISIBLE_MODULES\": \"0,1,2,3,4,5,6,7\",\n",
+    "            \"RAY_EXPERIMENTAL_NOSET_HABANA_VISIBLE_MODULES\": \"true\",\n",
```
Any particular reason to disable `RAY_EXPERIMENTAL_NOSET_HABANA_VISIBLE_MODULES`, and setup `WORLD_SIZE`, `HABANA_VISIBLE_MODULES` by yourself here? They should already be setup by Ray?

woshiyyya replied:
Can you check if these 3 env var settings can be removed?

harborn replied:
yes, removed these 3 env var.

woshiyyya commented on doc/source/train/examples/intel_gaudi/llama_deepspeed.ipynb:
```diff
@@ -0,0 +1,2436 @@
+   "source": [
+    "## Main Training Function\n",
+    "The `train_llama` function sets up the distributed training environment using Ray and starts the training process. To enable training using HPU, we only need to make the following changes:\n",
+    "- Set the exectuion mode for training, supported execution mode are:\n",
```
If you have tried these 3 modes, it would be a good idea to post the e2e training times here to show the performance of each mode.

woshiyyya replied:
Oh actually I'm referring to the performance of different execution modes. Like which one is faster? Does "eager.compile"> "lazy" > "eager"?

harborn replied:
"lazy" > "eager.compile" > "eager"

woshiyyya commented on doc/source/train/examples/intel_gaudi/llama_deepspeed.ipynb:
```diff
@@ -0,0 +1,2436 @@
+    "    # deepspeed config, can also place it to config file\n",
+    "    deepspeed_config = {\n",
+    "        \"steps_per_print\": 64,\n",
+    "        \"train_batch_size\": \"auto\",\n",
+    "        \"train_micro_batch_size_per_gpu\": \"auto\",\n",
+    "        \"gradient_accumulation_steps\": \"auto\",\n",
```
For curiosity, does `GaudiTrainer` or Deepspeed figure out the actual numbers at runtime?

harborn replied:
Yes, those option will be calculated when TrainingArguments object instanced.

woshiyyya reviewed (commented):
Thanks for the contributing these two high-quality examples @harborn ! Great demonstration of using LoRA and deepspeed ZERO-3 on Gaudi + Ray for fine-tuning.

Left some comments.

harborn commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
```diff
@@ -0,0 +1,831 @@
+    "[Habana Gaudi AI Processors (HPUs)](https://habana.ai) are AI hardware accelerators designed by Habana Labs. For more information, see [Gaudi Architecture](https://docs.habana.ai/en/latest/Gaudi_Overview/index.html) and [Gaudi Developer Docs](https://developer.habana.ai/).\n",
+    "\n",
+    "Basic features for this fine-tuning example are:\n",
+    "- Running on HPUs, support three execution mode: \"lazy\", \"eager\", \"eager.compile\".\n",
```
updated.

harborn commented on doc/source/train/examples.yml:
```diff
@@ -54,6 +54,26 @@ examples:
+    frameworks:
+      - accelerate
+      - transformers
+    skill_level: beginner
```
updated.

harborn commented on doc/source/train/examples.yml:
```diff
@@ -54,6 +54,26 @@ examples:
+    frameworks:
+      - accelerate
+      - transformers
+    skill_level: beginner
```
updated.

harborn commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
```diff
@@ -0,0 +1,831 @@
+    "    use_lazy_mode = True if execution_mode == \"lazy\" else False\n",
+    "    torch_compile_backend = \"hpu_backend\" if execution_mode == \"eager.compile\" else None\n",
+    "\n",
+    "    training_args = GaudiTrainingArguments(output_dir=config[\"output\"], do_train=True, do_eval=False, per_device_train_batch_size=config[\"batch_size_per_worker\"],\n",
```
OK, one argument one line?

woshiyyya replied:
yeah that'd be better

harborn commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
```diff
@@ -0,0 +1,831 @@
+   "metadata": {},
+   "source": [
+    "## Prepare environment\n",
+    "A node with Gaudi/Gaudi2 installed is required to run this example. This example will use 4 workers to train the model, each using 1 HPU.\n",
```
yes, will update

harborn commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
````diff
@@ -0,0 +1,831 @@
+    "```\n",
+    "### Install dependency\n",
+    "``` bash\n",
+    "pip install ray[train] notebook transformers datasets evaluate peft accelerate scikit-learn optimum-habana\n",
````
yes, updated

woshiyyya commented on doc/source/train/examples/intel_gaudi/llama_deepspeed.ipynb:
```diff
@@ -0,0 +1,2443 @@
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
```
```suggestion
  "orphan": true,
  "language_info": {
```

woshiyyya commented on doc/source/train/examples/intel_gaudi/llama.ipynb:
```diff
@@ -0,0 +1,849 @@
+   "language": "python",
+   "name": "python3"
+  },
+  "language_info": {
```
```suggestion
  "orphan": true,
  "language_info": {
```

woshiyyya commented:
@justinvyu Can you take a look and merge it?

justinvyu reviewed (commented):
The examples look good! Just a few requests:
1. Clear the cell outputs, and just put a mock markdown cell with the important output info. For example, just this information:
```
train_result = TrainOutput(global_step=62, training_loss=1.500297857869056, metrics={'train_runtime': 93.3311, 'train_samples_per_second': 71.042, 'train_steps_per_second': 2.222, 'total_flos': 4.02963202792489e+16, 'train_loss': 1.500297857869056, 'epoch': 2.0, 'memory_allocated (GB)': 34.51, 'max_memory_allocated (GB)': 78.72, 'total_memory_available (GB)': 94.62})
```
2. Is it possible to merge these two notebooks so that I can just flip a flag if I want to use deepspeed? Most of the logic is identical, just some extra configs.
3. (Just a question, not blocking) Should we also allow full parameter finetuning instead of always using lora?

harborn commented:
> The examples look good! Just a few requests:
> 
> 1. Clear the cell outputs, and just put a mock markdown cell with the important output info. For example, just this information:
> 
> ```
> train_result = TrainOutput(global_step=62, training_loss=1.500297857869056, metrics={'train_runtime': 93.3311, 'train_samples_per_second': 71.042, 'train_steps_per_second': 2.222, 'total_flos': 4.02963202792489e+16, 'train_loss': 1.500297857869056, 'epoch': 2.0, 'memory_allocated (GB)': 34.51, 'max_memory_allocated (GB)': 78.72, 'total_memory_available (GB)': 94.62})
> ```
> 
> 2. Is it possible to merge these two notebooks so that I can just flip a flag if I want to use deepspeed? Most of the logic is identical, just some extra configs.
> 3. (Just a question, not blocking) Should we also allow full parameter finetuning instead of always using lora?

Hi, Updated according your comments:
1. removed unnecessary cell outputs, only keep the important final outputs
2. merge two notebooks into one, and the final one notebook can run different training method on different execution mode on HPU.
3. Yes, if not using LoRA training, just remove LoRA conversion when loading pre-trained model.

harborn commented:
> The examples look good! Just a few requests:
> 
> 1. Clear the cell outputs, and just put a mock markdown cell with the important output info. For example, just this information:
> 
> ```
> train_result = TrainOutput(global_step=62, training_loss=1.500297857869056, metrics={'train_runtime': 93.3311, 'train_samples_per_second': 71.042, 'train_steps_per_second': 2.222, 'total_flos': 4.02963202792489e+16, 'train_loss': 1.500297857869056, 'epoch': 2.0, 'memory_allocated (GB)': 34.51, 'max_memory_allocated (GB)': 78.72, 'total_memory_available (GB)': 94.62})
> ```
> 
> 2. Is it possible to merge these two notebooks so that I can just flip a flag if I want to use deepspeed? Most of the logic is identical, just some extra configs.
> 3. (Just a question, not blocking) Should we also allow full parameter finetuning instead of always using lora?

woshiyyya commented:
@justinvyu Can you take a look again?

harborn commented:
> Seems that @harborn addressed the comments. @justinvyu could you take a look again?

@justinvyu please take a look again! Thanks.

justinvyu commented on doc/source/train/examples.yml:
```diff
@@ -54,6 +54,26 @@ examples:
+      - large language models
+    contributor: community
+    link: examples/intel_gaudi/llama
+  - title: "Fine-tune Llama-2-70b with Deepspeed on Intel Gaudi"
+    frameworks:
+      - accelerate
+      - transformers
+    skill_level: intermediate
+    use_cases:
+      - natural language processing
+      - large language models
+    contributor: community
+    link: examples/intel_gaudi/llama_deepspeed
```
```suggestion
```

harborn replied:
removed!

justinvyu commented on doc/source/train/examples.yml:
```diff
@@ -54,6 +54,26 @@ examples:
       - natural language processing
     contributor: community
     link: examples/intel_gaudi/bert
+  - title: "Fine-tune Llama-2-7b with Intel Gaudi"
+    frameworks:
+      - accelerate
+      - transformers
+    skill_level: intermediate
+    use_cases:
+      - natural language processing
+      - large language models
+    contributor: community
+    link: examples/intel_gaudi/llama
```
Can we move this to the end of the intermediate section in the yaml file, similar to the ordering of the beginner skill level sectoin?

I think it is better for the examples in each "skill level" to start with more familiar usage, then get more advanced into using custom accelerators.

harborn replied:
Updated!

justinvyu commented on doc/source/train/examples.yml:
```diff
@@ -54,6 +54,26 @@ examples:
       - natural language processing
     contributor: community
     link: examples/intel_gaudi/bert
+  - title: "Fine-tune Llama-2-7b with Intel Gaudi"
```
"Fine-tune Llama-2-7b and Llama-2-70b with Intel Gaudi"

harborn replied:
Updated!

justinvyu reviewed (commented):
Thanks for making the changes! One last comment then we can merge.

justinvyu commented:
@harborn I think you need to add back the `orphan: True` metadata: https://github.com/ray-project/ray/pull/44667#discussion_r1600456197

harborn commented:
> @harborn I think you need to add back the `orphan: True` metadata: [#44667 (comment)](https://github.com/ray-project/ray/pull/44667#discussion_r1600456197)

Fixed.
